File: stream-deliverer/src/streams/docker.rs
use std::fmt::Display;
use moq_async::Lock;

#[derive(Clone)]
pub struct Tag {
	pub name: String,
	pub version: String,
}

impl Display for Tag {
	fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
		write!(f, "{}:{}", self.name, self.version)
	}
}

#[derive(Clone)]
pub struct DockerFile {
	location: String,
	tag: Tag,
	context: String,
}
impl DockerFile {
	pub fn new(location: &str, tag: Tag, context: &str) -> Self {
		DockerFile {
			location: location.to_string(),
			tag,
			context: context.to_string(),
		}
	}

	pub fn get_location(&self) -> Location {
		self.location.clone()
	}
	pub fn get_tag(&self) -> Tag {
		self.tag.clone()
	}
	pub fn get_context(&self) -> Location {
		self.context.clone()
	}

	pub async fn build(&self) -> Image {
		const CMD: &str = format!(
			"docker build -t {} -f {} {}",
			self.tag.name, self.location, self.context
		)
		.as_str();

		tokio::process::Command::new(CMD)
			.await
			.expect("Failed to build docker image");

		Image::new(self.tag.clone())
	}
}

#[derive(Clone)]
pub struct Image {
	tag: Tag,
}
impl Image {
	pub fn new(tag: Tag) -> Self {
		Image { tag }
	}

	pub fn get_tag(&self) -> Tag {
		self.tag.clone()
	}
}

#[derive(Clone)]
pub enum Status {
	Running,
	Stopped,
	Exited,
}

impl PartialEq<Self> for Status {
	fn eq(&self, other: &Self) -> bool {
		match (self, other) {
			(Status::Running, Status::Running) => true,
			(Status::Stopped, Status::Stopped) => true,
			(Status::Exited, Status::Exited) => true,
			_ => false,
		}
	}
}

#[derive(Clone)]
pub struct Container {
	name: String,
	image: Image,
	ports: Vec<u16>,
	status: Lock<Status>,
}
impl Container {
	pub fn new(name: &str, image: Image) -> Self {
		Container {
			name: name.to_string(),
			image,
			ports: vec![],
			status: Lock::new(Status::Stopped),
		}
	}

	pub fn get_ports(&self) -> Vec<u16> {
		self.ports.clone()
	}
	pub fn get_name(&self) -> String {
		self.name.clone()
	}
	pub fn get_image(&self) -> Image {
		self.image.clone()
	}
	pub fn get_status(&self) -> Status {
		self.status.lock().clone()
	}

	pub fn add_port(&mut self, port: u16) {
		self.ports.push(port);
	}
	pub fn remove_port(&mut self, port: u16) {
		self.ports.retain(|&p| p != port);
	}

	pub async fn start(&self) -> Result<(), Err> {
		let mut status = self.status.lock();
		if *status == Status::Running {
			return Err("Container is already running".into());
		}
		*status = Status::Running;

		const CMD: &str = format!(
			"docker run --rm --name {} -p {} {}",
			self.name,
			self.ports
				.iter()
				.map(|port| format!("{}:{}", port, port))
				.collect::<Vec<String>>()
				.join(" "),
			self.image.get_tag()
		)
		.as_str();

		tokio::process::Command::new(CMD)
			.await
			.expect("Failed to start docker container");

		tracing::info!("Started container: {}", self.name);
	}

	pub async fn stop(&self) -> Result<(), Err> {
		let mut status = self.status.lock();
		if *status != Status::Running {
			return Err("Container is not running".into());
		}
		*status = Status::Stopped;

		const CMD: &str = format!("docker stop {}", self.name).as_str();
		tokio::process::Command::new(CMD)
			.await
			.expect("Failed to stop docker container");

		tracing::info!("Stopped container: {}", self.name);
	}
}

File: stream-deliverer/src/server/docker.ts
import { spawn } from 'child_process';

export class Tag {
    public name: string;
    public version: string;

    constructor(name: string, version: string) {
        this.name = name;
        this.version = version;
    }

    toString(): string {
        return `${this.name}:${this.version}`;
    }

    clone(): Tag {
        return new Tag(this.name, this.version);
    }
}

export class DockerFile {
    private location: string;
    private tag: Tag;
    private context: string;

    constructor(location: string, tag: Tag, context: string) {
        this.location = location;
        this.tag = tag;
        this.context = context;
    }

    getLocation(): string {
        return this.location;
    }

    getTag(): Tag {
        return this.tag.clone();
    }

    getContext(): string {
        return this.context;
    }

    async build(): Promise<Image> {
        const cmd = `docker build -t ${this.tag.name} -f ${this.location} ${this.context}`;

        return new Promise((resolve, reject) => {
            const process = spawn(cmd);

            process.on('error', (error) => {
                reject(new Error(`Failed to start docker build command: ${error.message}`));
            });

            process.on('close', (code) => {
                if (code === 0) {
                    resolve(new Image(this.tag.clone()));
                } else {
                    reject(new Error(`Docker build failed with exit code ${code}`));
                }
            });
        });
    }

    clone(): DockerFile {
        return new DockerFile(this.location, this.tag.clone(), this.context);
    }
}

export class Image {
    private tag: Tag;

    constructor(tag: Tag) {
        this.tag = tag;
    }

    getTag(): Tag {
        return this.tag.clone();
    }

    clone(): Image {
        return new Image(this.tag.clone());
    }
}

export enum Status {
    Running = "Running",
    Stopped = "Stopped",
    Exited = "Exited"
}

// Simple async lock implementation
class AsyncLock<T> {
    private value: T;
    private locked = false;
    private queue: Array<() => void> = [];

    constructor(initialValue: T) {
        this.value = initialValue;
    }

    async lock(): Promise<AsyncLockGuard<T>> {
        while (this.locked) {
            await new Promise<void>(resolve => {
                this.queue.push(resolve);
            });
        }
        this.locked = true;
        return new AsyncLockGuard(this);
    }

    unlock(): void {
        this.locked = false;
        const next = this.queue.shift();
        if (next) {
            next();
        }
    }

    getValue(): T {
        return this.value;
    }

    setValue(value: T): void {
        this.value = value;
    }
}

class AsyncLockGuard<T> {
    private lock: AsyncLock<T>;

    constructor(lock: AsyncLock<T>) {
        this.lock = lock;
    }

    get value(): T {
        return this.lock.getValue();
    }

    set value(val: T) {
        this.lock.setValue(val);
    }

    release(): void {
        this.lock.unlock();
    }
}

export class Container {
    private name: string;
    private image: Image;
    private ports: number[] = [];
    private status: AsyncLock<Status>;

    constructor(name: string, image: Image) {
        this.name = name;
        this.image = image;
        this.status = new AsyncLock(Status.Stopped);
    }

    getPorts(): number[] {
        return [...this.ports];
    }

    getName(): string {
        return this.name;
    }

    getImage(): Image {
        return this.image.clone();
    }

    async getStatus(): Promise<Status> {
        const guard = await this.status.lock();
        const status = guard.value;
        guard.release();
        return status;
    }

    addPort(port: number): void {
        this.ports.push(port);
    }

    removePort(port: number): void {
        this.ports = this.ports.filter(p => p !== port);
    }

    async start(): Promise<void> {
        const guard = await this.status.lock();

        if (guard.value === Status.Running) {
            guard.release();
            return;
        }

        guard.value = Status.Running;

        const portMappings = this.ports
            .map(port => `-p ${port}:${port}`)
            .join(' ');

        const cmd = `docker run --rm --name ${this.name} ${portMappings} ${this.image.getTag()}`;

        return new Promise((resolve, reject) => {
            const process = spawn(cmd);

            process.on('error', (error) => {
                guard.release();
                reject(new Error(`Failed to start docker run command: ${error.message}`));
            });

            process.on('spawn', () => {
                console.log(`Started container: ${this.name}`);
                guard.release();
                resolve();
            });
        });
    }

    async stop(): Promise<void> {
        const guard = await this.status.lock();

        if (guard.value !== Status.Running) {
            guard.release();
            return;
        }

        guard.value = Status.Stopped;

        const cmd = `docker stop ${this.name}`;

        return new Promise((resolve, reject) => {
            const process = spawn(cmd);

            process.on('error', (error) => {
                guard.release();
                reject(new Error(`Failed to start docker stop command: ${error.message}`));
            });

            process.on('close', (code) => {
                console.log(`Stopped container: ${this.name}`);
                guard.release();
                if (code === 0) {
                    resolve();
                } else {
                    reject(new Error(`Docker stop failed with exit code ${code}`));
                }
            });
        });
    }

    clone(): Container {
        const cloned = new Container(this.name, this.image.clone());
        cloned.ports = [...this.ports];
        return cloned;
    }
}